sys/net/link_layer/ieee802154/security.c

```diff
@@ -0,0 +1,528 @@
+    memcpy(tmp + sizeof(uint16_t), a, off);
+    _cbc_next(ctx, mic, tmp, tmp, sizeof(uint16_t) + off);
+    for (;off < a_len;) {
+        off += _cbc_next(ctx, mic, tmp, &(((uint8_t *)a)[off]), a_len - off);
+    }
+    for (off = 0; off < m_len;) {
+        off += _cbc_next(ctx, mic, tmp, &(((uint8_t *)m)[off]), m_len - off);
+    }
+}
+
+static void _ctr(ieee802154_sec_context_t *ctx,
+                 ieee802154_ccm_block_t *A0,
+                 const void *m, uint16_t m_len)
+{
+    uint8_t tmp1[IEEE802154_SEC_BLOCK_SIZE] = { 0 };
+    uint8_t tmp2[IEEE802154_SEC_BLOCK_SIZE] = { 0 };
+
+    for (uint16_t off = 0; off < m_len;) {
+        _advance_ctr_Ai(A0);
+        off += _ecb(ctx, tmp1, tmp2,
+                    &(((uint8_t *)m)[off]), (uint8_t *)A0, m_len - off);
+    }
+}
+
+static void _ctr_mic(ieee802154_sec_context_t *ctx,
+                     ieee802154_ccm_block_t *A0,
+                     void *mic, uint8_t mic_size)
+{
+    uint8_t tmp1[IEEE802154_SEC_BLOCK_SIZE] = { 0 };
+    uint8_t tmp2[IEEE802154_SEC_BLOCK_SIZE] = { 0 };
+
+    _ecb(ctx, tmp1, tmp2, mic, (uint8_t *)A0, mic_size);
+}
+
+void ieee802154_sec_init(ieee802154_sec_context_t *ctx)
+{
+    /* device driver can override this */
+    ctx->dev.cipher_ops = &ieee802154_radio_cipher_ops;
+    /* device driver can override this */
+    ctx->dev.ctx = ctx;
+    /* MIC64 is the only mandatory security mode */
+    ctx->security_level = IEEE802154_SCF_SECLEVEL_ENC_MIC64;
+    ctx->key_id_mode = IEEE802154_SCF_KEYMODE_IMPLICIT;
+    memset(ctx->key_source, 0, sizeof(ctx->key_source));
+    ctx->key_index = 0;
+    ctx->frame_counter = 0;
+    uint8_t key[] = IEEE802154_DEFAULT_KEY;
+
+    assert(CIPHER_MAX_CONTEXT_SIZE >= IEEE802154_SEC_KEY_LENGTH);
+    cipher_init(&ctx->cipher, CIPHER_AES_128, key, IEEE802154_SEC_KEY_LENGTH);
+}
+
+int ieee802154_sec_encrypt_frame(ieee802154_sec_context_t *ctx,
+                                 const uint8_t *header, uint8_t *header_size,
+                                 uint8_t *payload, uint16_t payload_size,
+                                 uint8_t *mic, uint8_t *mic_size,
+                                 const uint8_t *src_address)
+{
+    /* For non data frames (MAC commands, beacons) a and a_len would be larger.
+       ACKs are not encrypted. */
+    assert((*((uint8_t *)header)) & IEEE802154_FCF_TYPE_DATA);
+
+    if (ctx->security_level == IEEE802154_SCF_SECLEVEL_NONE) {
+        *mic_size = 0;
+        return IEEE802154_SEC_OK;
+    }
+    if (ctx->frame_counter == 0xFFFFFFFF) {
+        /* Letting the frame counter overflow is explicitly prohibited by the specification.
+           (see 9.4.2) */
+        return -IEEE802154_SEC_FRAME_COUNTER_OVERFLOW;
+    }
+    if (*header_size + payload_size >
+         /* TODO: support 802.15.4g frame sizes */
+        (IEEE802154_FRAME_LEN_MAX
+        - IEEE802154_MAX_AUX_HDR_LEN
+        - IEEE802154_FCS_LEN)) {
+        return -IEEE802154_SEC_FRAME_TOO_LARGE;
+    }
```

Comment:
Actually you can drop that check if you instead add 

```patch
--- a/drivers/netdev/ieee802154.c
+++ b/drivers/netdev/ieee802154.c
@@ -177,6 +177,9 @@ int netdev_ieee802154_get(netdev_ieee802154_t *dev, netopt_t opt, void *value,
 
             *((uint16_t *)value) = (_get_ieee802154_pdu(dev)
                                     - IEEE802154_MAX_HDR_LEN)
+#ifdef IEEE802154_SECURITY
+                                    - IEEE802154_MAX_AUX_HDR_LEN
+#endif
                                     - IEEE802154_FCS_LEN;
             res = sizeof(uint16_t);
             break;
```

That also solves the issue with 802.15.4g frames and prevents too large fragments from being generated in the first place.